[shawnnlimm] iP

shawnnlimm opened the pull request:
# Duke Chatbot

> "eat(); sleep(); code(); repeat();" - Anonymous Coder 

Duke chatbot is **EASY** to use and learn!. Some features include:
- Adding tasks
- Removing tasks
- Keeping track of deadlines

All you need to do is:
1. Download it from [here] (https://github.com/shawnnlimm/ip)
2. Run the jar file
3. That's it! 🔥 

Here's a checklist for you:
- [ ] Download jar file
- [x] Run jar file
- [ ] Complete!

If you already know Java, here's the `main` method:
``` java
public class Main {
    public static void main(String[] args) {
        Application.launch(MainApp.class, args);
    }
}
```

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

Commit:
docs/README.md: Tweak document template

Let's tweak the docs/README.md (which is used as the user guide)
to fit Duke better. Specifically,

1. mention product name in the title
2. mention adding a product screenshot and a product intro
3. tweak the flow to describe feature-by-feature

shawnnlimm committed:
feat: Add greetings

shawnnlimm committed:
feat: Add echo command

shawnnlimm committed:
feat: Add list feature

shawnnlimm committed:
feat: Add mark as done/undone feature

shawnnlimm committed:
feat: Add ToDo, Event, Deadline

shawnnlimm committed:
test: Add Automated Text UI Testing

shawnnlimm committed:
feat: Add error handling

shawnnlimm committed:
feat: Add Delete Task feature

shawnnlimm committed:
feat: Add Storage feature

shawnnlimm committed:
feat: Add Storage feature

shawnnlimm committed:
Revert "feat: Add Storage feature"

This reverts commit cb7afedc43aa8dd46d9809b892ba1a6355ae80fa.

shawnnlimm committed:
feat: Add Storage feature

shawnnlimm committed:
Merge branch 'branch-Level-7'

shawnnlimm committed:
feat: Add Date and Time feature

shawnnlimm committed:
Merge branch 'branch-Level-8'

shawnnlimm committed:
refactor: Add packages

shawnnlimm committed:
Merge branch 'add-gradle-support' of https://github.com/shawnnlimm/ip

shawnnlimm committed:
test: Add JUnit tests

shawnnlimm committed:
Merge branch 'branch-A-JUnit'

shawnnlimm committed:
docs: Add JavaDoc to non-private classes/methods

shawnnlimm committed:
Revert "docs: Add JavaDoc to non-private classes/methods"

This reverts commit ad6a74a76f3df8c397030f547a8eef2ac8270df6.

shawnnlimm committed:
docs: Add JavaDocs to non-private classes/methods

shawnnlimm committed:
style: Tweak code to comply with coding standards

shawnnlimm committed:
feat: Add find task with keyword feature

shawnnlimm committed:
fix: Fix merge conflicts

shawnnlimm committed:
Merge branch 'branch-Level-9'

shawnnlimm committed:
fix: Fix problem of jar file not working

shawnnlimm committed:
style: Fix checkstyle issues

shawnnlimm committed:
Merge branch 'branch-A-Checkstyle'

gingerbreaf commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,90 @@
+            String arguments = p.getArgument();
+
+            try {
+                switch (command) {
```
Would be more tidy if you made respective command classes instead.

gingerbreaf commented on src/main/java/duke/tasks/Deadline.java:
```diff
@@ -0,0 +1,47 @@
+ */
+public class Deadline extends Task {
+
+    private static final String DEADLINE_MESSAGE = "[D]%s (by: %s)";
+    private static final String DEADLINE_FILE_TEMPLATE = "D | %s | %s";
```
This is a good idea, wish i had thought of this

gingerbreaf commented on src/main/java/duke/tasks/Task.java:
```diff
@@ -0,0 +1,63 @@
+/**
+ * The Task class represents a generic task in the Duke chatbot application.
+ */
+public class Task {
```
Would it be better if this was an abstract class?

gingerbreaf commented on src/main/java/duke/tasks/TaskList.java:
```diff
@@ -0,0 +1,250 @@
+        taskList.add(deadline);
+        updateStorage();
+        String output = String.format(ADD_TASK_MESSAGE, deadline, taskList.size());
+        System.out.println(output);
```
I thought that these are meant to be done in the Ui class

gingerbreaf commented on src/main/java/duke/parser/Parser.java:
```diff
@@ -0,0 +1,79 @@
+        this.command = c;
+    }
+
+    /**
+     * Returns the parsed command.
+     *
+     * @return The parsed command.
+     */
```
The javadocs for these getters seems extra.

gingerbreaf commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,90 @@
+ * Duke chatbot.
+ */
+public class Duke {
+    private static final String GOODBYE_MESSAGE = "Bye. Hope to see you again soon!";
+    private static final String UNKNOWN_COMMAND_MESSAGE = "The command '%s' is unknown. Please try again!";
```
Good follow of coding naming standards

gingerbreaf commented on src/main/java/duke/tasks/Deadline.java:
```diff
@@ -0,0 +1,47 @@
+     *
+     * @return The string representation of the deadline task for saving to a file.
+     */
+    public String taskFileTemplate() {
```
I believe this should be toTaskFileTemplate. But good use of camelCase

gingerbreaf commented on src/main/java/duke/tasks/Event.java:
```diff
@@ -0,0 +1,51 @@
+     *
+     * @return The string representation of the event task for saving to a file.
+     */
+    public String taskFileTemplate() {
```
^

gingerbreaf commented on src/main/java/duke/tasks/TaskList.java:
```diff
@@ -0,0 +1,250 @@
+            "Noted. I've removed this task:%n%s%nNow you have %d tasks in the list.";
+    private static final String FIND_TASK_MESSAGE = "Here are the matching tasks in your list:%s";
+
+    private final ArrayList<Task> taskList;
+    private final Storage taskStorage;
```
Good use of camelCase

shawnnlimm committed:
feat: Add GUI to chatbot using JavaFX

shawnnlimm committed:
Merge branch 'branch-Level-10'

shawnnlimm committed:
style: Fix checkstyle issues

tituschewxj commented on src/main/java/duke/tasks/TaskList.java:
```diff
@@ -0,0 +1,248 @@
+     * @param toDoDescription The description of the ToDo task.
+     * @return String with return message.
+     * @throws DukeIllegalArgumentException If the description is black.
+     */
```
Perhaps you should include an empty line between the description and parameter section? I've noticed this pattern in several other places too.

tituschewxj commented on src/main/java/duke/tasks/TaskList.java:
```diff
@@ -0,0 +1,248 @@
+        }
+    }
+
+    private void syncListWithStorage(String taskType, String status, String description,
```
Perhaps this non-trival private method should carry header comments?

tituschewxj commented on src/main/java/duke/tasks/Event.java:
```diff
@@ -0,0 +1,51 @@
+ */
+public class Event extends Task {
+
+    private static final String EVENT_MESSAGE = "[E]%s (from: %s to: %s)";
+    private static final String EVENT_FILE_TEMPLATE = "E | %s | %s | %s";
```
I liked how you stored the template as constants. Perhaps the icon `E` can be stored as another constant as it is duplicated?

tituschewxj commented on src/main/java/duke/tasks/ToDo.java:
```diff
@@ -0,0 +1,39 @@
+        super(description);
+    }
+
+    /**
+     * Returns the string representation of the to-do task for saving to a file.
+     *
+     * @return The string representation of the to-do task for saving to a file.
+     */
```
Should `@return` be omitted, since the return value is obvious from the rest of the comment?

tituschewxj commented on src/main/java/duke/tasks/Task.java:
```diff
@@ -0,0 +1,63 @@
+     * @return The status icon.
+     */
+    public String getStatusIcon() {
+        return isDone ? "X" : " ";
```
Perhaps `"X"` can be a constant?

tituschewxj commented on src/main/java/duke/tasks/TaskList.java:
```diff
@@ -0,0 +1,248 @@
+        if (taskIndex.isBlank()) {
+            throw new DukeIllegalArgumentException(MISSING_ARGUMENT_MESSAGE);
+        }
+        Task taskToRemove = taskList.get(Integer.parseInt(taskIndex) - 1);
```
Should `Integer.parseInt(taskIndex)` be wrapped in a try-catch? Since it is possible for `taskIndex` to not be in the expected format, throwing a `NumberFormatException`?

tituschewxj commented on src/main/java/duke/storage/Storage.java:
```diff
@@ -0,0 +1,70 @@
+     * @param path The path to the file.
+     * @throws DukeIoException If an error occurs during file creation.
+     */
+    public Storage(String path) {
```
Perhaps your Constructor could explicitly declare using the `throws` clause in the method signature?
```suggestion
    public Storage(String path) throws DukeIoException {
```

tituschewxj commented on src/main/java/duke/tasks/TaskList.java:
```diff
@@ -0,0 +1,248 @@
+        StringBuilder sb = new StringBuilder();
+        int count = 1;
+        for (Task task : taskList) {
+            if (task.description.contains(keyword)) {
```
Perhaps there could be a method within `Task`, `doesDescriptionContains(String keyword)`, that checks if the task's description contains the keyword, instead of exposing the `description` outside of the abstraction barrier?

tituschewxj reviewed (commented):
Overall, I found your code easy to read, due to your good use of Javadocs comments. However, I felt that Javadocs may be a bit excessive at some parts. I noted that your code followed the coding standard well, and there are no major style violations.

shawnnlimm committed:
refactor: Add assertions for improved reliability

shawnnlimm committed:
refactor: Add assertions for improved reliability

shawnnlimm committed:
Revert "refactor: Add assertions for improved reliability"

This reverts commit 4aab2865304357bd36a8b8774d21be5ce982ef72.

shawnnlimm committed:
Revert "refactor: Add assertions for improved reliability"

This reverts commit b7f9ae2971eef586a2df5ded7a8eef2a0f472b53.

shawnnlimm committed:
refactor: Add assertions for improved reliability

shawnnlimm committed:
fix: Fix bug in DialogBox

The height of DialogBoxes were fixed.

This causes the list of tasks to be truncated when 'list' command is called.

By utilizing Region.USE_PREF_SIZE, the DialogBoxes heights are now variable.
This allows the full list of tasks to be shown correctly.

shawnnlimm committed:
Merge pull request #1 from shawnnlimm/branch-A-CodeQuality

fix: Fix bug in DialogBox

shawnnlimm committed:
feat: Implement A-CI

shawnnlimm committed:
feat: Implement feature to detect duplicate tasks

Currently, the chatbox allows users to input duplicate tasks.

This might cause the list contain unnecessary tasks and make it difficult
for users to find their tasks if the list is long.

Adding an error check for duplicates allows us to prevent users
from adding duplicate tasks.

Let's add a new class, DukeDuplicateException, that extends
DukeException. DukeDuplicateException will be thrown whenever a user
tries to add a duplicate task.

shawnnlimm committed:
Merge pull request #2 from shawnnlimm/branch-Extension

feat: Implement feature to detect duplicate tasks

shawnnlimm committed:
refactor: Renamed all instances of Duke to Waffles

shawnnlimm committed:
docs: Add User Guide

shawnnlimm committed:
fix: Fix bug for TaskListTest
